source/extensions/filters/http/cache/cache_filter.cc

```diff
@@ -129,23 +120,103 @@ Http::FilterDataStatus CacheFilter::encodeData(Buffer::Instance& data, bool end_
   return Http::FilterDataStatus::Continue;
 }
 
+void CacheFilter::getHeaders(Http::RequestHeaderMap& request_headers) {
+  ASSERT(lookup_, "CacheFilter is trying to call getHeaders with no LookupContext");
+
+  // If the cache posts a callback to the dispatcher then the CacheFilter is destroyed for any
+  // reason (e.g client disconnected and HTTP stream terminated), then there is no guarantee that
+  // the posted callback will run before the filter is deleted. Hence, a weak_ptr to the CacheFilter
+  // is captured and used to make sure the CacheFilter is still alive before accessing it in the
+  // posted callback.
+  // TODO(yosrym93): Look into other options for handling this (also in getBody and getTrailers) as
+  // they arise, e.g. cancellable posts, guaranteed ordering of posted callbacks and deletions, etc.
+  CacheFilterWeakPtr self = shared_from_this();
+
+  // The dispatcher needs to be captured because there's no guarantee that
+  // decoder_callbacks_->dispatcher() is thread-safe.
+  lookup_->getHeaders([self, &request_headers,
+                       &dispatcher = decoder_callbacks_->dispatcher()](LookupResult&& result) {
+    // The callback is posted to the dispatcher to make sure it is called on the worker thread.
+    // The lambda passed to dispatcher.post() needs to be copyable as it will be used to
+    // initialize a std::function. Therefore, it cannot capture anything non-copyable.
+    // LookupResult is non-copyable as LookupResult::headers_ is a unique_ptr, which is
+    // non-copyable. Hence, "result" is decomposed when captured, and re-instantiated inside the
+    // lambda so that "result.headers_" can be captured as a raw pointer, then wrapped in a
+    // unique_ptr when the result is re-instantiated.
+    dispatcher.post(
+        [self, &request_headers, status = result.cache_entry_status_,
+         headers = result.headers_.release(), response_ranges = std::move(result.response_ranges_),
+         content_length = result.content_length_, has_trailers = result.has_trailers_]() mutable {
+          if (auto cache_filter = self.lock()) {
+            cache_filter->onHeaders(LookupResult{status, absl::WrapUnique(headers), content_length,
+                                                 response_ranges, has_trailers},
+                                    request_headers);
+          }
```

Comment:
If this is false you are going to leak headers I think. You need to wrap it before the lock. Please add a test for this. You may want to consider thread_synchronizer if you want real threads, or you can do it with mocks.

Reply:
Good catch! Thanks for pointing this out.
The unit test I added in the last commit (in `cache_filter_test`) should be already exercising the case where the callback is made after the filter is deleted, which should be the case where the headers leak. Is there some kind of leak detector tests that run in the CI? What's Envoy's policy on testing for memory leaks?

Reply:
The heap checker runs by default and should check for leaks. I'm not sure why it's not catching this case but it should be.

Reply:
Is there a way to run it locally so that I can debug this?

Reply:
I wonder why it didn't catch this, I will take another look at the test.

Reply:
I fixed the leak but I don't know how the heap checker was not catching a memory leak before the fix. Maybe something is wrong with my test but I can't see it.

Reply:
Can you verify the leak checker works by leaking something else obvious, and then checking your test? I'm also concerned that your test is not testing what you think it is.

Reply:
Interesting observation, when I add a dummy leak like this:
```
dispatcher.post([self, &request_headers, status = result.cache_entry_status_,
                     headers_raw_ptr = result.headers_.release(),
                     response_ranges = std::move(result.response_ranges_),
                     content_length = result.content_length_,
                     has_trailers = result.has_trailers_,
                     dummy = new int]() mutable {
      // Wrap the raw pointer in a unique_ptr before checking to avoid memory leaks.
      if (CacheFilterSharedPtr cache_filter = self.lock()) {
        cache_filter->onHeaders(
            LookupResult{status, absl::WrapUnique(headers_raw_ptr), content_length, response_ranges, has_trailers},
            request_headers);
        absl::WrapUnique(dummy);
      }
    });
```

My test cause the heap checker to detect the leak of `dummy`, but not `headers_raw_ptr`.